### PR TITLE
warm up the query plan cache with persisted queries

### LINKED ISSUE #3770: add persisted queries to the query planner warm up
Persisted queries should be part of warmed up queries

### PR DESCRIPTION
**Pull request by Geal**
Fix #3770

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]
Just missing a changelog

**Comment by Geal** [comment]
I think we'll make one changelog entry for all the warmup improvements

File: apollo-router/src/query_planner/caching_query_planner.rs [hunk 1]
```diff
@@ -129,12 +131,36 @@ where
...
+        tracing::info!(
+            "warming up the query plan cache with {} queries, this might take a while",
+            cache_keys.len()
+                + persisted_queries_operations
+                    .as_ref()
+                    .map(|ops| ops.len())
+                    .unwrap_or(0)
+        );
+
+        let mut all_cache_keys = Vec::new();
```
Comment (bnjjj) [inline apollo-router/src/query_planner/caching_query_planner.rs:145]: can we initialize it with a capacity ? Like persisted_queries_operations.len() + cache_keys.len() ?

Comment (Geal) [reply apollo-router/src/query_planner/caching_query_planner.rs]: done in [5f3df65](https://github.com/apollographql/router/pull/3829/commits/5f3df656486ddf2d86fd641a955029c8e75f4865)

File: apollo-router/src/query_planner/caching_query_planner.rs [hunk 2]
```diff
@@ -129,12 +131,36 @@ where
...
+            for query in queries {
+                all_cache_keys.push(WarmUpCachingQueryKey {
+                    query,
+                    operation: None,
+                    metadata: CacheKeyMetadata::default(),
+                });
+            }
+        }
+
+        all_cache_keys.extend(cache_keys.into_iter());
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/query_planner/caching_query_planner.rs:159]: yes, that's the behaviour here. I added a comment in [7a3099a](https://github.com/apollographql/router/pull/3829/commits/7a3099ae34777ecdb2567d38a94680374622642e)

**Pull request merged**